tests/test_cli.py

```diff
@@ -443,3 +475,61 @@ def test_fail_if_invalid_env_provided_in_run(
             str(result.exception)
             == f"Invalid env-var: {env_var}, expected format: KEY=VALUE"
         )
+
+
+@pytest.mark.parametrize(
+    "on_job_scheduled",
+    (
+        "bad_str_format",
+        "nonexistant_module:func",
+        "tests.helpers.on_job_scheduled_helper:nonexistant_attr",
+        "tests.helpers.on_job_scheduled_helper:existing_attr",
+    ),
+    ids=(
+        "bad_str_format",
+        "no_module",
+        "no_attr",
+        "not_callable",
+    ),
+)
+def test_fail_if_invalid_on_job_scheduled_provided_in_run(
+    patched_kedro_package,
+    cli_context,
+    dummy_pipeline,
+    tmp_path: Path,
+    on_job_scheduled: str,
+):
+    create_kedro_conf_dirs(tmp_path)
+    with patch.dict(
+        "kedro.framework.project.pipelines", {"__default__": dummy_pipeline}
+    ), patch.object(Path, "cwd", return_value=tmp_path), patch(
+        "kedro_azureml.client.MLClient"
+    ) as ml_client_patched, patch(
+        "kedro_azureml.auth.utils.DefaultAzureCredential"
+    ), patch.dict(
+        os.environ, {"AZURE_STORAGE_ACCOUNT_KEY": "dummy_key"}
+    ):
+        # extra checks to make sure the tests actually checks the usecase
+        # and that we are not always in the case of an invalid string
+        if ":" in on_job_scheduled:
+            module_str, attr_str = on_job_scheduled.split(":")
```

Comment:
This does not test the code for `dynamic_import_job_schedule_func_from_str` at all 🤔 You're effectively validating the arguments passed to the test, not to the downstream invocation (`runner.invoke`). 
If you want to test this properly, the I would suggest:
1. Leave the cases as they are right now:
```python
    "on_job_scheduled",
    (
        "bad_str_format",
        "nonexistant_module:func",
        "tests.helpers.on_job_scheduled_helper:nonexistant_attr",
        "tests.helpers.on_job_scheduled_helper:existing_attr",
    ),
```
2. Run the `runner.invoke(...)`
3. Check the `result.exception` against one of the cases that you've specified:
```
assert result.exit_code != 0
assert result.exception, "Exception should have been raised"
# Rest of the checks
if module_str == "nonexistant_module":
    # ... rest of the checks based on the result.execption
```

Reply:
You're right, this is indeed a better solution.

My reasonning was that the only reason the test could fail was because of the --on-job-scheduled arg and I made sure to validate the arguments before the test. Also when checking the code coverage, the function was 100% covered so I figured it was enough.

But I agree with you it is not as robust as checking the error message after invoking the runner. Failure could happen because of something else. I don't know why I didn't think of that as it is the way it is done in another test 🤦‍♂️